File: python/tests/rootfinding_secant_test.py
from math import sqrt, cos
import pytest

from secant import secant


def test_rootfinding_secant_linear_function():
    def f(x):
        return x

    x0 = 1.0
    x1 = 2.0

    x = secant(f, x0, x1, 0.01)
    assert x == 0.0


def test_rootfinding_secant_linear_function_with_offset():
    def f(x):
        return x + 5.0

    x0 = 1.0
    x1 = 2.0

    x = secant(f, x0, x1, 0.01)
    assert x == -5.0


def test_rootfinding_secant_qudaratic_function():
    def f(x):
        return x * x

    x0 = 1.0
    x1 = 2.0

    delta = 0.00001

    x = secant(f, x0, x1, delta)
    assert abs(x - 0) < sqrt(delta)


def test_rootfinding_secant_qudaratic_function_with_offset():
    def f(x):
        return x * x - 1.0

    x0 = 3.0
    x1 = 2.0

    delta = 0.00001

    x = secant(f, x0, x1, delta)
    assert abs(x - 1.0) < delta


@pytest.mark.parametrize("x0, x1", [(3.0, 2.0),
                                    (2.0, 3.0),
                                    (0.1, 3.0),
                                    (2.5, 0.1),
                                    (0.5, 0.1)])
def test_rootfinding_secant_transcendent_function_with_starting_points(x0, x1):
    def f(x):
        return cos(x) - x

    delta = 0.00001
    expected_root = 0.739085
    calculated_root = secant(f, x0, x1, delta)

    assert abs(expected_root - calculated_root) < delta

File: cpp/tests/rootfinding_secant_test.cpp
#include "secant.hpp"
#include <gtest/gtest.h>
#include <tuple>

TEST(RootfindingSecantTest, LinearFunction)
{
    auto func = [](double x) { return x; };

    double const x0 = 1.0;
    double const x1 = 2.0;

    double const xprime = secant(func, x0, x1, 0.01);
    ASSERT_DOUBLE_EQ(0.0, xprime);
}

TEST(RootfindingSecantTest, LinearFunctionWithOffset)
{
    auto const func = [](double x) { return x + 5; };

    double const x0 = 1.0;
    double const x1 = 2.0;

    double const xprime = secant(func, x0, x1, 0.01);
    ASSERT_DOUBLE_EQ(-5.0, xprime);
}

TEST(RootfindingSecantTest, PolynomSecondOrder)
{
    auto const func = [](double x) { return x * x; };

    double const x0 = 1.0;
    double const x1 = 2.0;

    double const delta = 0.00001;

    double const xprime = secant(func, x0, x1, delta);
    ASSERT_LT(abs(xprime - 0), sqrt(delta));
}

TEST(RootfindingSecantTest, PolynomSecondOrderWithOffset)
{
    auto const func = [](double x) { return x * x - 1; };

    double const x0 = 3.0;
    double const x1 = 2.0;

    double const delta = 0.00001;

    double const xprime = secant(func, x0, x1, delta);
    ASSERT_LT(abs(xprime - 1.0), delta);
}

class RootfindingSecantWithStartParamTestFixture
    : public ::testing::TestWithParam<std::tuple<double, double>> {
};

TEST_P(RootfindingSecantWithStartParamTestFixture, TrascendentFunction)
{
    auto const func = [](double x) { return cos(x) - x; };

    double const x0 = std::get<0>(GetParam());
    double const x1 = std::get<1>(GetParam());

    double const delta = 0.00001;
    double const expected_root = 0.739085;

    double const calculated_root = secant(func, x0, x1, delta);
    ASSERT_LT(abs(calculated_root - expected_root), delta);
}

INSTANTIATE_TEST_SUITE_P(RootfindingSecantWithStartParamTest,
    RootfindingSecantWithStartParamTestFixture,
    ::testing::Values(std::make_tuple(3.0, 2.0), std::make_tuple(2.0, 3.0),
        std::make_tuple(0.1, 3.0), std::make_tuple(2.5, 0.1), std::make_tuple(0.5, 0.1)));
